File: android/app/src/test/java/com/nishany/android/feature/focus/FocusSessionTest.kt
package com.nishany.android.feature.focus

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

/** Mirrors `src/lib/focusSession.test.ts` case for case -- this is the one JVM test for the ported pure state machine; no UI is tested here. */
class FocusSessionTest {

    private val t0 = 1_700_000_000_000L

    @Test
    fun `start then tick then complete -- a countdown reaches zero and stops itself`() {
        val started = start(setDurationMinutes(initialFocusSession(t0), 5, t0), t0)
        assertEquals(true, started.running)
        assertEquals(300, started.remainingSeconds)

        val midway = tick(started, t0 + 150_000)
        assertEquals(true, midway.running)
        assertEquals(150, midway.remainingSeconds)
        assertEquals(150, accruedSeconds(midway))

        val finished = tick(midway, t0 + 300_000)
        assertEquals(false, finished.running)
        assertEquals(0, finished.remainingSeconds)
        assertEquals(t0 + 300_000, finished.completedAt)
        assertEquals(300, accruedSeconds(finished))
    }

    @Test
    fun `count-up mode accrues seconds instead of counting down`() {
        val started = start(setMode(initialFocusSession(t0), FocusMode.COUNTUP, t0), t0)
        val later = tick(started, t0 + 45_000)
        assertEquals(45, later.elapsedSeconds)
        assertEquals(45, accruedSeconds(later))
        assertNull(later.completedAt)
    }

    @Test
    fun `pause freezes the clock -- resuming with start picks up from there`() {
        val started = start(setDurationMinutes(initialFocusSession(t0), 25, t0), t0)
        val ran = tick(started, t0 + 10_000)
        val paused = pause(ran, t0 + 10_000)
        assertEquals(false, paused.running)
        assertEquals(25 * 60 - 10, paused.remainingSeconds)

        // Time passing while paused must not be counted.
        val stillPaused = tick(paused, t0 + 60_000)
        assertEquals(paused.remainingSeconds, stillPaused.remainingSeconds)

        val resumed = start(stillPaused, t0 + 60_000)
        assertEquals(true, resumed.running)
        assertEquals(paused.remainingSeconds, resumed.remainingSeconds)
    }

    @Test
    fun `starting again after completion restarts the block rather than staying at 0-00`() {
        val finished = tick(start(setDurationMinutes(initialFocusSession(t0), 5, t0), t0), t0 + 300_000)
        assertEquals(0, finished.remainingSeconds)
        val restarted = start(finished, t0 + 300_000)
        assertEquals(true, restarted.running)
        assertEquals(300, restarted.remainingSeconds)
        assertNull(restarted.completedAt)
    }

    @Test
    fun `strict-discard -- leaving mid-block throws the running progress away`() {
        val started = start(setDurationMinutes(initialFocusSession(t0), 25, t0), t0)
        val ran = tick(started, t0 + 5 * 60_000)
        assertEquals(20 * 60, ran.remainingSeconds)

        val discarded = discard(ran, t0 + 5 * 60_000 + 15_000)
        assertEquals(false, discarded.running)
        assertEquals(25 * 60, discarded.remainingSeconds)
        assertEquals(0, accruedSeconds(discarded))
        assertNull(discarded.completedAt)
    }

    @Test
    fun `reset returns to the full chosen duration in countdown, and zero in count-up`() {
        val countdown = reset(tick(start(setDurationMinutes(initialFocusSession(t0), 50, t0), t0), t0 + 5_000), t0 + 5_000)
        assertEquals(50 * 60, countdown.remainingSeconds)

        val countup = reset(tick(start(setMode(initialFocusSession(t0), FocusMode.COUNTUP, t0), t0), t0 + 5_000), t0 + 5_000)
        assertEquals(0, countup.elapsedSeconds)
    }

    @Test
    fun `formatClock prints mm-ss under an hour and h-mm-ss past it`() {
        assertEquals("1:05", formatClock(65))
        assertEquals("1:01:01", formatClock(3_661))
        assertEquals("0:00", formatClock(-4))
    }
}

File: src/lib/focusSession.test.ts
import test from 'node:test'
import assert from 'node:assert/strict'
import {
  accruedSeconds, discard, formatClock, initialFocusSession, pause, reset, setDurationMinutes,
  setMode, start, tick,
} from './focusSession.ts'

const T0 = 1_700_000_000_000

test('start -> tick -> complete: a countdown reaches zero and stops itself', () => {
  // The shortest duration the picker allows (`MIN_DURATION_MINUTES`) is 5 —
  // `setDurationMinutes` clamps anything under that, so the block below is
  // 5 minutes, not literally one.
  const started = start(setDurationMinutes(initialFocusSession(T0), 5, T0), T0)
  assert.equal(started.running, true)
  assert.equal(started.remainingSeconds, 300)

  const midway = tick(started, T0 + 150_000)
  assert.equal(midway.running, true)
  assert.equal(midway.remainingSeconds, 150)
  assert.equal(accruedSeconds(midway), 150)

  const finished = tick(midway, T0 + 300_000)
  assert.equal(finished.running, false)
  assert.equal(finished.remainingSeconds, 0)
  assert.equal(finished.completedAt, T0 + 300_000)
  assert.equal(accruedSeconds(finished), 300)
})

test('count-up mode accrues seconds instead of counting down', () => {
  const started = start(setMode(initialFocusSession(T0), 'countup', T0), T0)
  const later = tick(started, T0 + 45_000)
  assert.equal(later.elapsedSeconds, 45)
  assert.equal(accruedSeconds(later), 45)
  assert.equal(later.completedAt, null)
})

test('pause freezes the clock; resuming with start picks up from there', () => {
  const started = start(setDurationMinutes(initialFocusSession(T0), 25, T0), T0)
  const ran = tick(started, T0 + 10_000)
  const paused = pause(ran, T0 + 10_000)
  assert.equal(paused.running, false)
  assert.equal(paused.remainingSeconds, 25 * 60 - 10)

  // Time passing while paused must not be counted.
  const stillPaused = tick(paused, T0 + 60_000)
  assert.equal(stillPaused.remainingSeconds, paused.remainingSeconds)

  const resumed = start(stillPaused, T0 + 60_000)
  assert.equal(resumed.running, true)
  assert.equal(resumed.remainingSeconds, paused.remainingSeconds)
})

test('starting again after completion restarts the block rather than staying at 0:00', () => {
  const finished = tick(start(setDurationMinutes(initialFocusSession(T0), 5, T0), T0), T0 + 300_000)
  assert.equal(finished.remainingSeconds, 0)
  const restarted = start(finished, T0 + 300_000)
  assert.equal(restarted.running, true)
  assert.equal(restarted.remainingSeconds, 300)
  assert.equal(restarted.completedAt, null)
})

test('strict-discard: leaving mid-block throws the running progress away', () => {
  const started = start(setDurationMinutes(initialFocusSession(T0), 25, T0), T0)
  const ran = tick(started, T0 + 5 * 60_000)
  assert.equal(ran.remainingSeconds, 20 * 60)

  const discarded = discard(ran, T0 + 5 * 60_000 + 15_000)
  assert.equal(discarded.running, false)
  assert.equal(discarded.remainingSeconds, 25 * 60)
  assert.equal(accruedSeconds(discarded), 0)
  assert.equal(discarded.completedAt, null)
})

test('reset returns to the full chosen duration in countdown, and zero in count-up', () => {
  const countdown = reset(tick(start(setDurationMinutes(initialFocusSession(T0), 50, T0), T0), T0 + 5_000), T0 + 5_000)
  assert.equal(countdown.remainingSeconds, 50 * 60)

  const countup = reset(tick(start(setMode(initialFocusSession(T0), 'countup', T0), T0), T0 + 5_000), T0 + 5_000)
  assert.equal(countup.elapsedSeconds, 0)
})

test('formatClock prints mm:ss under an hour and h:mm:ss past it', () => {
  assert.equal(formatClock(65), '1:05')
  assert.equal(formatClock(3_661), '1:01:01')
  assert.equal(formatClock(-4), '0:00')
})
